File: app/src/main/java/xyz/rodit/snapmod/PathManager.java
package xyz.rodit.snapmod;

import android.net.Uri;
import android.os.Build;
import android.os.Environment;

import java.io.File;
import java.util.HashMap;
import java.util.Map;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import xyz.rodit.xposed.client.ConfigurationClient;

public class PathManager {

    public static final String DOWNLOAD_STORY = "story";
    public static final String DOWNLOAD_AUDIO_NOTE = "audio_note";
    public static final String DOWNLOAD_PROFILE = "profile";

    private static final Pattern PATTERN_PUBLIC_DIR = Pattern.compile("\\$(\\w*)");
    private static final Pattern PATTERN_PARAMETER = Pattern.compile("%(\\w)");

    private static final Map<String, String> publicDirs = new HashMap<>();
    private static final Map<String, String> defaultPaths = new HashMap<>();

    static {
        publicDirs.put("Movies", Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_MOVIES).getPath());
        publicDirs.put("Pictures", Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_PICTURES).getPath());
        publicDirs.put("Alarms", Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_ALARMS).getPath());
        publicDirs.put("DCIM", Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_DCIM).getPath());
        publicDirs.put("Music", Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_MUSIC).getPath());
        publicDirs.put("Ringtones", Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_RINGTONES).getPath());

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
            publicDirs.put("Recordings", Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_RECORDINGS).getPath());
            publicDirs.put("Screenshots", Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_SCREENSHOTS).getPath());
        }

        defaultPaths.put(DOWNLOAD_STORY, "$Movies/SnapMod/%u_story_%t");
        defaultPaths.put(DOWNLOAD_AUDIO_NOTE, "$Movies/SnapMod/%id_audio_%t");
        defaultPaths.put(DOWNLOAD_PROFILE, "$Movies/SnapMod/%u_profile_%t");
    }

    private static Map<String, String> appendDefaultParamsMap(Map<String, String> paramsMap) {
        Map<String, String> appended = new HashMap<>(paramsMap);
        appended.put("t", String.valueOf(System.currentTimeMillis()));
        return appended;
    }

    public static Map<String, String> createParams(String... params) {
        Map<String, String> paramsMap = new HashMap<>();
        for (int i = 0; i < params.length; i += 2) {
            paramsMap.put(params[i], params[i + 1]);
        }

        return paramsMap;
    }

    public static File getPath(ConfigurationClient config, String downloadType, Map<String, String> paramsMap, String extension) {
        Map<String, String> params = appendDefaultParamsMap(paramsMap);
        String path = config.getString("download_path_" + downloadType, defaultPaths.get(downloadType));
        Matcher matcher = PATTERN_PUBLIC_DIR.matcher(path);
        if (matcher.find()) {
            String replacement = publicDirs.get(matcher.group(1));
            if (replacement != null) {
                path = matcher.replaceFirst(replacement);
            }
        }

        matcher = PATTERN_PARAMETER.matcher(path);
        StringBuffer buffer = new StringBuffer();
        while (matcher.find()) {
            String replacement = params.get(matcher.group(1));
            if (replacement != null) {
                matcher.appendReplacement(buffer, replacement);
            }
        }

        matcher.appendTail(buffer);
        return new File(buffer + extension);
    }

    public static String getUri(ConfigurationClient config, String downloadType, Map<String, String> paramsMap, String extension) {
        return Uri.fromFile(getPath(config, downloadType, paramsMap, extension)).toString();
    }
}

File: app/src/main/java/xyz/rodit/snapmod/util/PathManager.kt
package xyz.rodit.snapmod.util

import android.net.Uri
import android.os.Build
import android.os.Environment
import xyz.rodit.xposed.client.ConfigurationClient
import java.io.File
import java.text.SimpleDateFormat
import java.util.*
import java.util.regex.Pattern

object PathManager {
    const val DOWNLOAD_STORY = "story"
    const val DOWNLOAD_AUDIO_NOTE = "audio_note"
    const val DOWNLOAD_PROFILE = "profile"
    const val DOWNLOAD_SNAP = "snap"

    private const val DEFAULT_DATE_FORMAT = "dd-MM-yyyy_HH-mm-ss"

    private val PATTERN_PUBLIC_DIR = Pattern.compile("""\$(\w+)""")
    private val PATTERN_PARAMETER = Pattern.compile("%([A-Za-z]+)")
    private val publicDirs: MutableMap<String, String> = mutableMapOf(
        "Movies" to Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_MOVIES).path,
        "Pictures" to Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_PICTURES).path,
        "Alarms" to Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_ALARMS).path,
        "DCIM" to Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_DCIM).path,
        "Music" to Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_MUSIC).path,
        "Ringtones" to Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_RINGTONES).path
    )
    private val defaultPaths: MutableMap<String, String> = mutableMapOf(
        DOWNLOAD_STORY to "\$Movies/SnapMod/%u_story_%d",
        DOWNLOAD_AUDIO_NOTE to "\$Movies/SnapMod/%u_audio_%d",
        DOWNLOAD_PROFILE to "\$Movies/SnapMod/%u_profile_%d",
        DOWNLOAD_SNAP to "\$Movies/SnapMod/%u_snap_%d"
    )

    private fun appendDefaultParamsMap(config: ConfigurationClient, paramsMap: Map<String, String>): Map<String, String> {
        val appended: MutableMap<String, String> = HashMap(paramsMap)
        appended["t"] = System.currentTimeMillis().toString()
        val dateFormat =
            SimpleDateFormat(
                config.getString("download_date_format", DEFAULT_DATE_FORMAT),
                Locale.getDefault()
            )
        appended["d"] = dateFormat.format(Date())
        return appended
    }

    fun getPath(
        config: ConfigurationClient,
        downloadType: String,
        paramsMap: Map<String, String>,
        extension: String
    ): File {
        val params = appendDefaultParamsMap(config, paramsMap)
        var path = config.getString("download_path_$downloadType", defaultPaths[downloadType])

        var matcher = PATTERN_PUBLIC_DIR.matcher(path)
        if (matcher.find()) {
            publicDirs[matcher.group(1).orEmpty()]?.let {
                path = matcher.replaceFirst(it)
            }
        }

        matcher = PATTERN_PARAMETER.matcher(path)
        val buffer = StringBuffer()
        while (matcher.find()) {
            params[matcher.group(1).orEmpty()]?.let {
                matcher.appendReplacement(buffer, it)
            }
        }
        matcher.appendTail(buffer)

        return File(buffer.toString() + extension)
    }

    fun getUri(
        config: ConfigurationClient,
        downloadType: String,
        paramsMap: Map<String, String>,
        extension: String
    ): String {
        return Uri.fromFile(getPath(config, downloadType, paramsMap, extension)).toString()
    }

    init {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
            publicDirs["Recordings"] =
                Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_RECORDINGS).path
            publicDirs["Screenshots"] =
                Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_SCREENSHOTS).path
        }
    }
}